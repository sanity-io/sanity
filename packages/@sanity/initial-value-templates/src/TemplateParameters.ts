export type TemplateParameter = FieldDefinition | ArrayFieldDefinition

export interface TypeTarget {
  type: string
}

export interface ReferenceTarget {
  type: 'reference'
  to: TypeTarget | TypeTarget[]
}

export interface FieldDefinition {
  name: string
  type: string
  title?: string
  description?: string
  options?: {[key: string]: any}
}

export type ReferenceFieldDefinition = FieldDefinition & ReferenceTarget

export type ArrayFieldDefinition = FieldDefinition & {
  type: 'array'
  of: (ReferenceTarget | TypeTarget)[]
}
