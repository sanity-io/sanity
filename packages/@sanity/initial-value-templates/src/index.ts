export {default as TemplateBuilder} from './builder'
export {resolveInitialValue, isBuilder} from './resolve'
export {validateTemplates} from './validate'
export {
  templateExists,
  getTemplates,
  getTemplateById,
  getTemplatesBySchemaType,
  getParameterlessTemplatesBySchemaType,
  getTemplateErrors
} from './templates'
