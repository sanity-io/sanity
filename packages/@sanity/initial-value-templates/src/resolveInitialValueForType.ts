import {isEmpty, resolveTypeName} from '@sanity/util/content'

import {
  ArraySchemaType,
  InitialValueParams,
  isArraySchemaType,
  isObjectSchemaType,
  ObjectSchemaType,
  SchemaType,
} from '@sanity/types'

import {randomKey} from '@sanity/util/paths'
import deepAssign from './util/deepAssign'
import {resolveValue} from './util/resolveValue'

export function getItemType(arrayType: ArraySchemaType, item: unknown): SchemaType | undefined {
  const itemTypeName = resolveTypeName(item)

  return itemTypeName === 'object' && arrayType.of.length === 1
    ? arrayType.of[0]
    : arrayType.of.find((memberType) => memberType.name === itemTypeName)
}

const MAX_RECURSION_DEPTH = 10

/**
 * Resolve initial value for the given schema type (recursively)
 *
 * @param type {SchemaType} this is the name of the document
 * @param params {Record<string, unknown>} params is a sanity context object passed to every initial value function
 * @param maxDepth {Record<string, unknown>} maximum recursion depth (default 9)
 */
export function resolveInitialValueForType(
  type: SchemaType,
  params: InitialValueParams = {},
  maxDepth = MAX_RECURSION_DEPTH
) {
  if (maxDepth <= 0) {
    return undefined
  }
  if (isObjectSchemaType(type)) {
    return resolveInitialObjectValue(type, params, maxDepth)
  }
  if (isArraySchemaType(type)) {
    return resolveInitialArrayValue(type, params, maxDepth)
  }
  return resolveValue(type.initialValue, params)
}

async function resolveInitialArrayValue(type, params: InitialValueParams, maxDepth: number) {
  const initialArray = await resolveValue(type.initialValue)
  return Array.isArray(initialArray)
    ? Promise.all(
        initialArray.map(async (initialItem) => {
          const itemType = getItemType(type, initialItem)!
          return isObjectSchemaType(itemType)
            ? {
                ...initialItem,
                ...(await resolveInitialValueForType(itemType, params, maxDepth - 1)),
                _key: randomKey(),
              }
            : initialItem
        })
      )
    : undefined
}
export async function resolveInitialObjectValue(
  type: ObjectSchemaType,
  params: InitialValueParams,
  maxDepth: number
) {
  const initialObject: Record<string, unknown> = {
    ...((await resolveValue(type.initialValue, params)) || {}),
  }

  const fieldValues = {}
  await Promise.all(
    type.fields.map(async (field) => {
      const initialFieldValue = await resolveInitialValueForType(field.type, params, maxDepth - 1)
      if (initialFieldValue !== undefined && initialFieldValue !== null) {
        fieldValues[field.name] = initialFieldValue
      }
    })
  )

  const merged = deepAssign(fieldValues, initialObject)
  if (isEmpty(merged)) {
    return undefined
  }
  if (type.name !== 'object') {
    merged._type = type.name
  }
  return merged
}
