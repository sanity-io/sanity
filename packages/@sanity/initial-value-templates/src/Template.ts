import {TemplateParameter} from './TemplateParameters'

type ValueResolver = (parameters: {[key: string]: any}) => {[key: string]: any}

export interface Template {
  id: string
  title: string
  description?: string
  schemaType: string
  icon?: Function
  value: ValueResolver | {[key: string]: any}
  parameters?: TemplateParameter[]
}

export class TemplateBuilder {
  spec: Partial<Template>

  constructor(spec?: Template) {
    this.spec = spec || {}
  }

  id(id: string) {
    return this.clone({id})
  }

  getId() {
    return this.spec.id
  }

  title(title: string) {
    return this.clone({title})
  }

  description(description: string) {
    return this.clone({description})
  }

  getDescription() {
    return this.spec.description
  }

  getTitle() {
    return this.spec.title
  }

  schemaType(typeName: string) {
    return this.clone({schemaType: typeName})
  }

  getSchemaType() {
    return this.spec.schemaType
  }

  icon(icon: Function) {
    return this.clone({icon})
  }

  getIcon() {
    return this.spec.icon
  }

  value(value: ValueResolver | {[key: string]: any}) {
    return this.clone({value})
  }

  getValue() {
    return this.spec.value
  }

  parameters(parameters: TemplateParameter[]) {
    return this.clone({parameters})
  }

  getParameters() {
    return this.spec.parameters
  }

  serialize(): Template {
    const {id, title, description, schemaType, value, icon, parameters} = this.spec
    if (!id) {
      throw new Error('Template is missing required "id"')
    }

    if (!title) {
      throw new Error(`Template with ID "${id}" is missing required "title"`)
    }

    if (!schemaType) {
      throw new Error(`Template with ID "${id}" is missing required "schemaType"`)
    }

    if (!value) {
      throw new Error(`Template with ID "${id}" is missing required "value"`)
    }

    return {id, title, description, schemaType, value, icon, parameters}
  }

  clone(withSpec?: Partial<Template>) {
    const builder = new TemplateBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}
    return builder
  }
}
