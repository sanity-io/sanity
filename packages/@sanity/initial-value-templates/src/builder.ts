import {Template, TemplateBuilder} from './Template'
import {Schema, SchemaType, getDefaultSchema} from './parts/Schema'

function defaultTemplateForType(
  schemaType: string | SchemaType,
  sanitySchema?: Schema
): TemplateBuilder {
  let type: SchemaType
  if (typeof schemaType === 'string') {
    const schema = sanitySchema || getDefaultSchema()
    type = schema.get(schemaType)
  } else {
    type = schemaType
  }

  return new TemplateBuilder({
    id: type.name,
    schemaType: type.name,
    title: type.title || type.name,
    icon: type.icon,
    value: type.initialValue || {_type: type.name},
  })
}

function defaults(sanitySchema?: Schema) {
  const schema = sanitySchema || getDefaultSchema()
  if (!schema) {
    throw new Error(
      'Unable to automatically resolve schema. Pass schema explicitly: `defaults(schema)`'
    )
  }

  return schema
    .getTypeNames()
    .filter((typeName) => !/^sanity\./.test(typeName))
    .filter((typeName) => isDocumentSchemaType(typeName, schema))
    .map((typeName) => defaultTemplateForType(schema.get(typeName), schema))
}

function isDocumentSchemaType(typeName: string, schema: Schema) {
  const schemaType = schema.get(typeName)
  return schemaType.type && schemaType.type.name === 'document'
}

export default {
  template: (spec?: Template) => new TemplateBuilder(spec),
  defaults,
  defaultTemplateForType,
}
