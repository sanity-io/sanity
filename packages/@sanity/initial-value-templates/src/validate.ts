import oneline from 'oneline'
import {isPlainObject} from 'lodash'
import {randomKey, toString as pathToString} from '@sanity/util/paths'
import {Template} from './Template'
import {TemplateParameter} from './TemplateParameters'
import {getDefaultSchema} from './parts/Schema'

export {validateInitialValue, validateTemplates}

const ALLOWED_REF_PROPS = ['_key', '_ref', '_weak', '_type']
const REQUIRED_TEMPLATE_PROPS: (keyof Template)[] = ['id', 'title', 'schemaType', 'value']

function validateTemplates(templates: Template[]) {
  const idMap = new Map()

  templates.forEach((template, i) => {
    const id = templateId(template, i)

    if (typeof (template as {[key: string]: any}).values !== 'undefined' && !template.value) {
      throw new Error(`Template ${id} is missing "value" property, but contained "values". Typo?`)
    }

    const missing = REQUIRED_TEMPLATE_PROPS.filter(prop => !template[prop])
    if (missing.length > 0) {
      throw new Error(`Template ${id} is missing required properties: ${missing.join(', ')}`)
    }

    if (typeof template.value !== 'function' && !isPlainObject(template.value)) {
      throw new Error(
        `Template ${id} has an invalid "value" property; should be a function or an object`
      )
    }

    if (typeof template.parameters !== 'undefined') {
      if (Array.isArray(template.parameters)) {
        template.parameters.forEach((param, i) => validateParameter(param, template, i))
      } else {
        throw new Error(`Template ${id} has an invalid "parameters" property; must be an array`)
      }
    }

    if (idMap.has(template.id)) {
      const dupeIndex = idMap.get(template.id)
      const dupe = `${quote(templates[dupeIndex].title)} at index ${dupeIndex}`
      throw new Error(
        `Template "${template.title}" at index ${i} has the same ID ("${template.id}") as template ${dupe}`
      )
    }

    idMap.set(template.id, i)
  })

  return templates
}

function templateId(template: Template, i: number) {
  return quote(template.id || template.title) || (typeof i === 'number' && `at index ${i}`) || ''
}

function quote(str: string) {
  return str && str.length > 0 ? `"${str}"` : str
}

function validateInitialValue(value: {[key: string]: any}, template: Template) {
  const contextError = (msg: string) => `Template "${template.id}" initial value: ${msg}`

  if (!isPlainObject(value)) {
    throw new Error(contextError(`resolved to a non-object`))
  }

  if (value._type && template.schemaType !== value._type) {
    throw new Error(
      contextError(oneline`
        includes "_type"-property (${value._type})
        that does not match template (${template.schemaType})
      `)
    )
  }

  try {
    return validateValue(value)
  } catch (err) {
    err.message = contextError(err.message)
    throw err
  }
}

function validateValue(value: any, path: (string | number)[] = [], parentIsArray = false): any {
  if (Array.isArray(value)) {
    return value.map((item, i) => {
      if (Array.isArray(item)) {
        throw new Error(
          `multidimensional arrays are not supported (at path "${pathToString(path)}")`
        )
      }

      return validateValue(item, path.concat(i), true)
    })
  }

  if (!isPlainObject(value)) {
    return value
  }

  // Apply missing keys is the parent is an array
  const initial: {[key: string]: any} = parentIsArray && !value._key ? {_key: randomKey()} : {}

  // Ensure non-root objects have _type
  if (path.length > 0 && !value._type) {
    if (value._ref) {
      // In the case of references, we know what the type should be, so apply it
      initial._type = 'reference'
    } else {
      throw new Error(`missing "_type" property at path "${pathToString(path)}"`)
    }
  }

  if (value._ref) {
    validateReference(value, path)
  }

  // Validate deeply
  return Object.keys(value).reduce((acc, key) => {
    acc[key] = validateValue(value[key], path.concat([key]))
    return acc
  }, initial)
}

function validateParameter(parameter: TemplateParameter, template: Template, index: number) {
  const schema = getDefaultSchema()

  if (!parameter.name) {
    throw new Error(
      `Template ${template.id} has a parameter at index ${index} that is missing its "name"-property`
    )
  }

  // I know, this is a weird one
  if (parameter.name === 'template') {
    throw new Error(
      `Template parameters cannot be named "template", see parameter #${index} for template ${template.id}`
    )
  }

  if (!schema.get(parameter.type)) {
    throw new Error(
      `Template parameter "${parameter.name}" has an invalid/unknown type: "${parameter.type}"`
    )
  }
}

function validateReference(value, path: (string | number)[] = []) {
  if (!value._type && value.type) {
    throw new Error(
      `Reference is missing "_type", but has a "type" property at path "${pathToString(path)}"`
    )
  }

  const disallowed = Object.keys(value).filter(key => !ALLOWED_REF_PROPS.includes(key))
  if (disallowed.length > 0) {
    const plural = disallowed.length > 1 ? 'properties' : 'property'
    throw new Error(
      `Disallowed ${plural} found in reference: ${disallowed
        .map(quote)
        .join(', ')} at path "${pathToString(path)}"`
    )
  }
}
