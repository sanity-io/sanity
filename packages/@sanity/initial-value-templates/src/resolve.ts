import {isPlainObject} from 'lodash'
import {Template, TemplateBuilder} from './Template'
import {validateInitialValue} from './validate'

export function isBuilder(template: Template | TemplateBuilder): template is TemplateBuilder {
  return typeof (template as TemplateBuilder).serialize === 'function'
}

async function resolveInitialValue(
  template: Template | TemplateBuilder,
  params: {[key: string]: any} = {}
): Promise<{[key: string]: any}> {
  // Template builder?
  if (isBuilder(template)) {
    return resolveInitialValue(template.serialize(), params)
  }

  const {id, value} = template
  if (!value) {
    throw new Error(`Template "${id}" has invalid "value" property`)
  }

  // Static value?
  if (isPlainObject(value)) {
    return validateInitialValue(value, template)
  }

  // Not an object, so should be a function
  if (typeof value !== 'function') {
    throw new Error(
      `Template "${id}" has invalid "value" property - must be a plain object or a resolver function`
    )
  }

  const resolved = await value(params)
  return validateInitialValue(resolved, template)
}

export {resolveInitialValue}
