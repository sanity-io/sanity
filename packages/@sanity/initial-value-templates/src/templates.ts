/* eslint-disable import/prefer-default-export */
import T from './builder'
import {validateTemplates} from './validate'
import {Template, TemplateBuilder} from './Template'
import {Schema} from './parts/Schema'
import {isBuilder} from './resolve'

interface CommonJsEsStub {
  __esModule: boolean
  default: (Template | TemplateBuilder)[]
}

function isCommonJsEsStub(mod: any): mod is CommonJsEsStub {
  const stub = mod as CommonJsEsStub
  return stub.__esModule && 'default' in stub
}

function loadDefinedTemplatesFromPart() {
  try {
    const templates:
      | undefined
      | (Template | TemplateBuilder)[]
      | CommonJsEsStub = require('part:@sanity/base/initial-value-templates?')

    if (typeof templates === 'undefined') {
      return undefined
    }

    if (!templates) {
      return null
    }

    if (isCommonJsEsStub(templates)) {
      return templates.default
    }

    return templates
  } catch (err) {
    return undefined
  }
}

function loadDefinedTemplates(schema?: Schema) {
  let templates = loadDefinedTemplatesFromPart()

  // Templates is `null` if the part was implemented but the export doesn't make sense,
  // if it is not found (require failed), the value will be `undefined`
  if (templates === null || (templates && !Array.isArray(templates))) {
    const type = templates === null ? `null` : typeof templates
    throw new Error(
      `'part:@sanity/base/initial-value-templates' should be an array of templates, got ${type}`
    )
  }

  if (!templates) {
    templates = T.defaults(schema)
  }

  return prepareTemplates(templates)
}

function prepareTemplates(templates: (Template | TemplateBuilder)[]) {
  const serialized = templates.map(maybeSerialize)
  return validateTemplates(serialized)
}

function getDefaultTemplates(schema?: Schema) {
  return prepareTemplates(T.defaults(schema))
}

function maybeSerialize(template: Template | TemplateBuilder) {
  return isBuilder(template) ? template.serialize() : template
}

export function getTemplateErrors(schema: Schema) {
  try {
    loadDefinedTemplates(schema)
    return []
  } catch (err) {
    return [err]
  }
}

export function getTemplates(schema?: Schema) {
  try {
    return loadDefinedTemplates(schema)
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn('Failed to load defined templates, falling back to defaults:\n%s', err.message)
    return getDefaultTemplates(schema)
  }
}

export function getParameterlessTemplatesBySchemaType(schemaType: string) {
  return getTemplatesBySchemaType(schemaType).filter(
    tpl => !tpl.parameters || !tpl.parameters.length
  )
}

export function getTemplatesBySchemaType(schemaType: string) {
  return getTemplates().filter(tpl => tpl.schemaType === schemaType)
}

export function getTemplateById(id: string) {
  return getTemplates().find(tpl => tpl.id === id)
}

export function templateExists(id: string) {
  return Boolean(getTemplateById(id))
}
