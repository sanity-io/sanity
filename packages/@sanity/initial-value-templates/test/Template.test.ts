import {TemplateBuilder as T} from '../src'

const icon = () => null

describe('T.template()', () => {
  test('builds template through constructor', () => {
    expect(
      T.template({
        id: 'foo',
        title: 'some title',
        schemaType: 'author',
        value: {name: 'Default name!'},
        icon
      }).serialize()
    ).toMatchSnapshot()
  })

  test('throws on missing id', () => {
    expect(() => T.template().serialize()).toThrowError(/required "id"/)
  })

  test('throws on missing title', () => {
    expect(() =>
      T.template()
        .id('id')
        .serialize()
    ).toThrowError(/required "title"/)
  })

  test('throws on missing schemaType', () => {
    expect(() =>
      T.template()
        .id('id')
        .title('Blah')
        .serialize()
    ).toThrowError(/required "schemaType"/)
  })

  test('throws on missing value', () => {
    expect(() =>
      T.template()
        .id('id')
        .title('Blah')
        .schemaType('author')
        .serialize()
    ).toThrowError(/required "value"/)
  })

  test('can construct using builder', () => {
    expect(
      T.template()
        .id('yeah')
        .title('Yeah')
        .schemaType('author')
        .icon(icon)
        .value({name: 'bar'})
        .serialize()
    ).toMatchSnapshot()
  })

  test('builder is immutable', () => {
    const original = T.template()
    expect(original.id('foo')).not.toEqual(original)
    expect(original.title('foo')).not.toEqual(original)
    expect(original.schemaType('author')).not.toEqual(original)
    expect(original.icon(icon)).not.toEqual(original)
    expect(original.value({name: 'foo'})).not.toEqual(original)
  })

  test('getters work', () => {
    const original = T.template()
    expect(original.id('foo').getId()).toEqual('foo')
    expect(original.title('bar').getTitle()).toEqual('bar')
    expect(original.schemaType('author').getSchemaType()).toEqual('author')
    expect(original.icon(icon).getIcon()).toEqual(icon)
    expect(original.value({name: 'bar'}).getValue()).toEqual({name: 'bar'})
  })
})

describe('T.defaultTemplateForType()', () => {
  test('generates correct representation of type without initial value', () => {
    expect(T.defaultTemplateForType('author').serialize()).toMatchSnapshot()
  })

  test('generates correct representation of type with initial value', () => {
    expect(T.defaultTemplateForType('post').serialize()).toMatchSnapshot()
  })
})

describe('T.defaults()', () => {
  test('generates array of all schema type templates', () => {
    expect(T.defaults()).toMatchSnapshot()
    expect(T.defaults().map(tpl => tpl.serialize())).toMatchSnapshot()
  })
})
