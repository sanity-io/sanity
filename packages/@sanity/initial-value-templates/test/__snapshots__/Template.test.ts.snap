// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`T.defaultTemplateForType() generates correct representation of type with initial value 1`] = `
Object {
  "description": undefined,
  "icon": [Function],
  "id": "post",
  "parameters": undefined,
  "schemaType": "post",
  "title": "Post",
  "value": Object {
    "_type": "post",
  },
}
`;

exports[`T.defaultTemplateForType() generates correct representation of type without initial value 1`] = `
Object {
  "description": undefined,
  "icon": [Function],
  "id": "author",
  "parameters": undefined,
  "schemaType": "author",
  "title": "Author",
  "value": [Function],
}
`;

exports[`T.defaults() generates array of all schema type templates 1`] = `
Array [
  TemplateBuilder {
    "spec": Object {
      "icon": [Function],
      "id": "author",
      "schemaType": "author",
      "title": "Author",
      "value": [Function],
    },
  },
  TemplateBuilder {
    "spec": Object {
      "icon": [Function],
      "id": "post",
      "schemaType": "post",
      "title": "Post",
      "value": Object {
        "_type": "post",
      },
    },
  },
]
`;

exports[`T.defaults() generates array of all schema type templates 2`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`T.template() builds template through constructor 1`] = `
Object {
  "description": undefined,
  "icon": [Function],
  "id": "foo",
  "parameters": undefined,
  "schemaType": "author",
  "title": "some title",
  "value": Object {
    "name": "Default name!",
  },
}
`;

exports[`T.template() can construct using builder 1`] = `
Object {
  "description": undefined,
  "icon": [Function],
  "id": "yeah",
  "parameters": undefined,
  "schemaType": "author",
  "title": "Yeah",
  "value": Object {
    "name": "bar",
  },
}
`;
